---
title: "Iterative single-cell multi-omic integration using online learning"
author: "Chao Gao and Joshua Welch"
date: "Dec 18, 2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Scenario 1: sampling minibatches from fully observed datasets
To perform online iNMF, we need to install the `online` branch. Please see the instruction below.
```{r install, message=FALSE, results='hide', eval=FALSE}
library(devtools)
install_github("welch-lab/liger")
```

We first create a liger object by passing the filenames of HDF5 files containing the raw count data. The data can be downloaded [here](https://umich.box.com/s/m5cojytshxvk44pc4kki6clh6c9sv7ai). Liger assumes by default that the HDF5 files are formatted by the 10X CellRanger pipeline. Large datasets are often generated over multiple 10X runs (for example, multiple biological replicates). In such cases it may be necessary to merge the HDF5 files from each run into a single HDF5 file. We provide the `mergeH5` function for this purpose (see below for details).

```{r create.object, message=FALSE, results='hide'}
library(liger)
pbmcs = createLiger(list(stim = "pbmcs_stim.h5", ctrl = "pbmcs_ctrl.h5"))
```

We then perform the normalization, gene selection, and gene scaling in an online fashion, reading the data from disk in small batches.
```{r preprocess, message=FALSE, results='hide'}
pbmcs = normalize(pbmcs)
pbmcs = selectGenes(pbmcs, var.thresh = 0.2, do.plot = F)
pbmcs = scaleNotCenter(pbmcs)
```
## Online Integrative Nonnegative Matrix Factorization
Now we can use online iNMF to factorize the data, again using only minibatches that we read from the HDF5 files on demand (default mini-batch size = 5000). Sufficient number of iterations is crucial for obtaining ideal factorization result. If the size of the mini-batch is set to be close to the size of the whole dataset (i.e. an epoch only contains one iteration), `max.epochs` needs to be increased accordingly for more iterations. 
```{r factorize, message=FALSE, results='hide'}
pbmcs = online_iNMF(pbmcs, k = 20, miniBatch_size = 5000, max.epochs = 5)
```
## Quantile Normalization and Downstream Analysis
After performing the factorization, we can perform quantile normalization to align the datasets.
```{r norm, message=FALSE, results='hide'}
pbmcs = quantile_norm(pbmcs)
```

We can also visualize the cell factor loadings in two dimensions using t-SNE or UMAP.
```{r visualize}
pbmcs = runUMAP(pbmcs)
plotByDatasetAndCluster(pbmcs, axis.labels = c("UMAP1","UMAP2"))
```

Let's first evaluate the quality of data alignment. The alignment score ranges from 0 (no alignment) to 1 (perfect alignment).
```{r alignmentScore}
calcAlignment(pbmcs)
```

With HDF5 files as input, we need to sample the raw, normalized, or scaled data from the full dataset on disk and load them in memory. Some plotting functions and downstream analyses are designed to operate on a subset of cells sampled from the full dataset. This enables rapid analysis using limited memory. The readSubset function allows either uniform random sampling or sampling balanced by cluster. Here we extract the normalized count data of 5000 sampled cells.
```{r subset, results='hide'}
pbmcs = readSubset(pbmcs, slot.use = "norm.data", max.cells = 5000)
```

Using the sampled data stored in memory, we can now compare clusters or datasets (within each cluster) to identify differentially expressed genes. The runWilcoxon function
performs differential expression analysis by performing an in-memory Wilcoxon rank-sum test on this subset. Thus, users can still analyze large datasets with a fixed amount of memory.
```{r wilcox 1, results='hide'}
de_genes = runWilcoxon(pbmcs, compare.method = "datasets")
```

Here we show the top 10 genes in cluster 1 whose expression level significantly differ between two dataset.
```{r wilcox 2}
de_genes = de_genes[order(de_genes$padj), ]
head(de_genes[de_genes$group == "1-stim",], n = 10)
```

We can show UMAP coordinates of sampled cells by their loadings on each factor (Factor 1 as an example). Underneath it displays the
most highly loading shared and dataset-specific genes, with the size of the marker indicating the magnitude of the loading.
```{r wordClouds, warning=FALSE, results='hide', fig.keep = 'last'}
p_wordClouds = plotWordClouds(pbmcs, num.genes = 5, return.plots = T)
p_wordClouds[[1]]
```

We can generate plots of dimensional reduction coordinates colored by expression of specified gene. The first two UMAP dimensions and gene ISG15 (identified by Wilcoxon test in the previous step) is used as an example here.
```{r gene}
plotGene(pbmcs, "ISG15", return.plots = F)
```

Furthermore, we can make violin plots of expression of specified gene for each dataset (ISG15 as an example).
```{r violin, warning=FALSE}
plotGeneViolin(pbmcs, "ISG15", return.plots = F)
```

The online algorithm can be implemented on datasets loaded in memory as well. The same analysis is performed on the PBMCs, shown below.  
```{r , message=FALSE, results='hide'}
stim = readRDS("pbmcs_stim.RDS")
ctrl = readRDS("pbmcs_ctrl.RDS")
pbmcs_mem = createLiger(list(stim = stim, ctrl = ctrl), remove.missing = F)
pbmcs_mem = normalize(pbmcs_mem)
pbmcs_mem = selectGenes(pbmcs_mem, var.thresh = 0.2, do.plot = F)
pbmcs_mem = scaleNotCenter(pbmcs_mem)
pbmcs_mem = online_iNMF(pbmcs_mem, k = 20, miniBatch_size = 5000, max.epochs = 5)
```

```{r ,message=FALSE}
pbmcs_mem = quantile_norm(pbmcs_mem)
pbmcs_mem = runUMAP(pbmcs_mem)
plotByDatasetAndCluster(pbmcs_mem, axis.labels = c("UMAP1","UMAP2"))
```

As mentioned above, it is sometimes necessary to merge multiple HDF5 files (such as multiple 10X runs from the same tissue or condition) into a single file. We provide the `mergeH5` function for this purpose. The function takes as input a list of filenames to merge (file.list), a vector of sample names that are prepended to the cell barcodes (library.names), and the name of the merged HDF5 file. The function requires that all files to be merged include exactly the same set of genes. For example, we can merge the cells and nuclei datasets used in the examples below (note that merging these particular two datasets is not something you would like to do, and is purely to demonstrate the `mergeH5` function).
```{r merge,message=FALSE}
mergeH5(file.list = list("allen_smarter_cells.h5", "allen_smarter_nuclei.h5"), 
        library.names = c("cells","nuclei"),
        new.filename = "cells_nuclei")
```

## Scenario 2: iterative refinement by incorporating new datasets
We can also perform online iNMF with continually arriving datasets.
```{r prep1, message=FALSE, results='hide'}
MOp = createLiger(list(cells = "allen_smarter_cells.h5"))
MOp = normalize(MOp)
MOp = selectGenes(MOp, var.thresh = 2)
MOp.vargenes = MOp@var.genes
MOp = scaleNotCenter(MOp)
```

```{r fact1, message=FALSE, results='hide'}
MOp = online_iNMF(MOp, k = 40, max.epochs = 1)
```

```{r vis1}
MOp = quantile_norm(MOp)
MOp = runUMAP(MOp)
plotByDatasetAndCluster(MOp, axis.labels = c("UMAP1","UMAP2"))
```

```{r prep2, message=FALSE, results='hide'}
MOp2 = createLiger(list(nuclei = "allen_smarter_nuclei.h5"))
MOp2 = normalize(MOp2)
MOp2@var.genes = MOp@var.genes
MOp2 = scaleNotCenter(MOp2)
```

```{r fact2, message=FALSE, results='hide'}
MOp = online_iNMF(MOp, X_new = list(nuclei = MOp2), k = 40, max.epochs = 1)
```

```{r vis2}
MOp = quantile_norm(MOp)
MOp = runUMAP(MOp)
plotByDatasetAndCluster(MOp, axis.labels = c("UMAP1","UMAP2"))
```

## Scenario 3: projecting new datasets
```{r prep3, message=FALSE, results='hide'}
MOp = createLiger(list(cells = "allen_smarter_cells.h5"))
MOp@var.genes = MOp.vargenes
MOp = online_iNMF(MOp, k = 40, max.epochs = 1)
MOp = quantile_norm(MOp)
MOp = runUMAP(MOp)
plotByDatasetAndCluster(MOp, axis.labels = c("UMAP1","UMAP2"))
```

```{r project, message=FALSE, results='hide'}
MOp = online_iNMF(MOp, X_new = list(nuclei = MOp2), k = 40, project = TRUE)
```

```{r vis3}
MOp = quantile_norm(MOp)
MOp = runUMAP(MOp)
plotByDatasetAndCluster(MOp, axis.labels = c("UMAP1","UMAP2"))
```








