---
title: "Unshared Features UINMF"
author: "Joshua Welch and April Kriebel"
date: "12/03/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
osmFISH = readRDS("C:/Users/april/OneDrive/Documents/UINMF_Final_backups/Vignette/OSMFISH.vin.RDS")
rna = readRDS("C:/Users/april/OneDrive/Documents/UINMF_Final_backups/Vignette/DROPVIZ.vin.RDS")
library(rliger)

```
```{r eval = FALSE}
install.packages('devtools')
library(devtools)
install_github('welch-lab/liger')
```

# Step 1: Preprocessing and Normalization

First, read in your datasets. For this tutorial, we will use the osmFISH dataset (33 genes by 5,219 cells), and a downsampled single-cell RNA-seq dataset (29,463 genes by 15,000 cells). The datasets can be downloaded at https://www.dropbox.com/sh/y9kjoum8u469nj1/AADik2b2-Qo3os2QSWXdIAbna?dl=0

```{r eval = FALSE}
osmFISH = readRDS("OSMFISH.vin.RDS")
rna = readRDS("DROPVIZ.vin.RDS")
```

Next, create your Liger object, submitting the datasets in list format. The unshared features should not be subsetted out, or submitted separately. Rather, they should be included in the matrix submitted for that dataset. For example,the scRNA-seq data is submitted in its entirety, the unshared features are not submitted separately. This helps ensure proper normalization.
```{r createliger, warning = FALSE}
osm.liger <- createLiger(list(osmFISH = osmFISH, rna = rna))
```
Normalize the datasets.The normalization is applied to the datasets in their entirety.

```{r normalize}
osm.liger <- normalize(osm.liger)
```
To include unshared features in your analysis, set the unshared parameter to TRUE when selecting genes.When selecting the unshared features, it is necessary to include a list of what datasets unshared features should be included for. For instance, in this case, we wish to include the unshared features from the RNA dataset, the second dataset in our analysis. We provide an individual tuning threshold for unshared features. If a single value is submitted, that threshold is applied to all datasets' unshared features. If multiple datasets include unshared features, the user can specify an individual thresholds for each dataset by submitting a list of thresholds the same length as the number of datasets with unshared datasets.
The variable unshared features willl be stored in `liger@var.unshared.features`.

```{r selectGenes, warning= FALSE}
osm.liger <- selectGenes(osm.liger, unshared = TRUE, unshared.datasets = list(2), unshared.thresh= 0.4)
```
The scaleNotCenter functions will scale both the shared and unshared features. The scaled unshared features will be stored 
in `liger@scale.unshared.data`

```{r scalenotcenter}
osm.liger <- scaleNotCenter(osm.liger)
```
# Step 2: Joint Matrix Factorization

To factorize the datasets including unshared datasets, set the use.unshared parameter to TRUE. 


```{r factorization}
osm.liger <- optimizeALS(osm.liger, k=30, use.unshared = TRUE)
```
# Step 3: Quantile Normalization and Joint Clustering

After factorization, the resulting Liger object can used in all downstream LIGER functions without adjustment. The default reference dataset for quantile normalization is the larger dataset, but the user should select the higher quality dataset as the reference dataset, even if it is the smaller dataset.


```{r quantilenorm, results = FALSE}
osm.liger <- quantile_norm(osm.liger, ref_dataset= "rna")
osm.liger <- louvainCluster(osm.liger)
```
# Step 4: Visualizations and Downstream processing

```{r runumap, results = FALSE, warnings = FALSE, message = FALSE}
osm.liger <- runUMAP(osm.liger)
```
Next, we can visualize our returned factorized object by dataset to check the alignment between datasets, as well as by cluster determined in the factorization.
```{r visualizations, message = FALSE, warning = FALSE}
umap_plots <-plotByDatasetAndCluster(osm.liger, axis.labels = c("UMAP1","UMAP2"), return.plots = TRUE)
umap_plots[[1]]
umap_plots[[2]]
```
\
We can also examine features such as gene expression. 
```{r gene visualizations}
Pdgfra <- plotGene(osm.liger, "Pdgfra", axis.labels = c('UMAP 1', 'UMAP 2'), return.plots = TRUE)
Bmp4 <- plotGene(osm.liger, "Bmp4", axis.labels = c('UMAP 1', 'UMAP 2'), return.plots = TRUE)
plot_grid(Pdgfra[[2]],Pdgfra[[1]],Bmp4[[2]],Bmp4[[1]], ncol=2)
```


