<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Linked Inference of Genomic Experimental Relationships • rliger2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="Linked Inference of Genomic Experimental Relationships">
<meta property="og:description" content="Uses an extension of nonnegative matrix factorization to identify shared and dataset-specific factors. See Welch J, Kozareva V, et al (2019) &lt;doi:10.1016/j.cell.2019.05.006&gt;, and Liu J, Gao C, Sodicoff J, et al (2020) &lt;doi:10.1038/s41596-020-0391-8&gt; for more details.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown-header">Installation</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    </li>
    <li>
      <a href="articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    </li>
    <li>
      <a href="articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    </li>
    <li>
      <a href="articles/UINMF_vignette.html">Integration using unshared features</a>
    </li>
    <li>
      <a href="articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    </li>
    <li>
      <a href="articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    </li>
    <li>
      <a href="articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    </li>
    <li>
      <a href="articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    </li>
    <li>
      <a href="articles/Parameter_selection.html">Parameter selection</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Usage
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Import data</li>
    <li class="dropdown-header">Introduction to a liger object</li>
    <li class="dropdown-header">Interoperate with Seurat</li>
  </ul>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">NEWS</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<p><img src="https://github.com/mvfki/liger/raw/newObj/inst/extdata/logo.png" width="120"></p>
<p><a href="https://github.com/mvfki/liger/actions/workflows/r.yml" class="external-link"><img src="https://github.com/mvfki/liger/actions/workflows/r.yml/badge.svg?branch=newObj" alt="R"></a><a href="https://codecov.io/gh/mvfki/liger" class="external-link"><img src="https://codecov.io/gh/mvfki/liger/branch/newObj/graph/badge.svg?token=77TTU4GY8A" alt="codecov"></a></p>
<div class="section level1">
<div class="page-header"><h1 id="liger-linked-inference-of-genomic-experimental-relationships">LIGER (Linked Inference of Genomic Experimental Relationships)<a class="anchor" aria-label="anchor" href="#liger-linked-inference-of-genomic-experimental-relationships"></a>
</h1></div>
<p>LIGER (installed as <code>rliger</code> ) is a package for integrating and analyzing multiple single-cell datasets, developed by the Macosko lab and maintained/extended by the Welch lab. It relies on integrative non-negative matrix factorization to identify shared and dataset-specific factors.</p>
<p>Check out our <a href="https://doi.org/10.1016/j.cell.2019.05.006" class="external-link">Cell paper</a> for a more complete description of the methods and analyses. To access data used in our SN and BNST analyses, visit our <a href="https://portals.broadinstitute.org/single_cell/study/SCP466" class="external-link">study</a> on the Single Cell Portal.</p>
<p>LIGER can be used to compare and contrast experimental datasets in a variety of contexts, for instance:</p>
<ul>
<li>Across experimental batches</li>
<li>Across individuals</li>
<li>Across sex</li>
<li>Across tissues</li>
<li>Across species (e.g., mouse and human)</li>
<li>Across modalities (e.g., scRNAseq and spatial transcriptomics data, scMethylation, or scATAC-seq)</li>
</ul>
<p>Once multiple datasets are integrated, the package provides functionality for further data exploration, analysis, and visualization. Users can:</p>
<ul>
<li>Identify clusters</li>
<li>Find significant shared (and dataset-specific) gene markers</li>
<li>Compare clusters with previously identified cell types</li>
<li>Visualize clusters and gene expression using t-SNE and UMAP</li>
</ul>
<p>We have also designed LIGER to interface with existing single-cell analysis packages, including <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a>.</p>
<div class="section level2">
<h2 id="feedback">Feedback<a class="anchor" aria-label="anchor" href="#feedback"></a>
</h2>
<p>Consider filling out our <a href="https://forms.gle/bhvp3K6tiHwf976r8" class="external-link">feedback form</a> to help us improve the functionality and accessibility of LIGER.</p>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>For usage examples and guided walkthroughs, check the <code>vignettes</code> directory of the repo.</p>
<ul>
<li><p><a href="http://htmlpreview.github.io/?https://github.com/welch-lab/liger/blob/master/vignettes/online_iNMF_tutorial.html" class="external-link">Iterative Single-Cell Multi-Omic Integration Using Online iNMF</a></p></li>
<li><p><a href="http://htmlpreview.github.io/?https://github.com/welch-lab/liger/blob/master/vignettes/UINMF_vignette.html" class="external-link">Integrating unshared features with UINMF</a></p></li>
<li><p><a href="http://htmlpreview.github.io/?https://github.com/welch-lab/liger/blob/master/vignettes/SNAREseq_walkthrough.html" class="external-link">scATAC and scRNA Integration using unshared features (UINMF)</a></p></li>
<li><p><a href="http://htmlpreview.github.io/?https://github.com/welch-lab/liger/blob/master/vignettes/cross_species_vig.html" class="external-link">Cross-species Analysis with UINMF</a></p></li>
<li><p><a href="http://htmlpreview.github.io/?https://github.com/welch-lab/liger/blob/master/vignettes/Parameter_selection.html" class="external-link">Performing Parameter Selection</a></p></li>
<li><p><a href="https://www.dropbox.com/s/ho62az1kshj59d2/STARmap_dropviz_vig.html?dl=0" class="external-link">Integrating spatial transcriptomic and transcriptomic datasets using UINMF</a> (Click to Download)</p></li>
<li><p><a href="http://htmlpreview.github.io/?https://github.com/welch-lab/liger/blob/master/vignettes/Integrating_multi_scRNA_data.html" class="external-link">Integrating Multiple Single-Cell RNA-seq Datasets</a></p></li>
<li><p><a href="http://htmlpreview.github.io/?https://github.com/welch-lab/liger/blob/master/vignettes/Integrating_scRNA_and_scATAC_data.html" class="external-link">Jointly Defining Cell Types from scRNA-seq and scATAC-seq</a></p></li>
<li><p><a href="https://welch-lab.github.io/liger/rna-methylation.html" class="external-link">Jointly Defining Cell Types from Single-Cell RNA-seq and DNA Methylation</a></p></li>
<li><p><a href="https://htmlpreview.github.io/?https://github.com/satijalab/seurat.wrappers/blob/master/docs/liger.html" class="external-link">Running Liger directly on Seurat objects using Seurat wrappers</a></p></li>
</ul>
</div>
<div class="section level2">
<h2 id="system-requirements">System Requirements<a class="anchor" aria-label="anchor" href="#system-requirements"></a>
</h2>
<div class="section level3">
<h3 id="hardware-requirements">Hardware requirements<a class="anchor" aria-label="anchor" href="#hardware-requirements"></a>
</h3>
<p>The <code>rliger</code> package requires only a standard computer with enough RAM to support the in-memory operations. For minimal performance, please make sure that the computer has at least about 2 GB of RAM. For optimal performance, we recommend a computer with the following specs:</p>
<ul>
<li>RAM: 16+ GB</li>
<li>CPU: 4+ cores, 2.3 GHz/core</li>
</ul>
</div>
<div class="section level3">
<h3 id="software-requirements">Software requirements<a class="anchor" aria-label="anchor" href="#software-requirements"></a>
</h3>
<p>The package development version is tested on <em>Linux</em> operating systems and <em>Mac OSX</em>.</p>
<ul>
<li>Linux: CentOS 7, Manjaro 5.3.18</li>
<li>Mac OSX: Mojave (10.14.1), Catalina (10.15.2)</li>
</ul>
<p>The <code>rliger</code> package should be compatible with Windows, Mac, and Linux operating systems.</p>
<p>Before setting up the <code>rliger</code> package, users should have R version 3.4.0 or higher, and several packages set up from CRAN and other repositories. The user can check the dependencies in <code>DESCRIPTION</code>.</p>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>LIGER is written in R and is also available on the Comprehensive R Archive Network (CRAN). Note that the package name is <code>rliger</code> to avoid a naming conflict with an unrelated package. To install the version on CRAN, follow these instructions:</p>
<ol style="list-style-type: decimal">
<li>Install <a href="https://www.r-project.org/" class="external-link">R</a> (&gt;= 3.6)</li>
<li>Install <a href="https://posit.co/download/rstudio-desktop/" class="external-link">Rstudio</a> (recommended)</li>
<li>Type the following R command:</li>
</ol>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'rliger'</span><span class="op">)</span></span></code></pre>
<p>To install the latest development version directly from GitHub, type the following commands instead of step 3:</p>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'devtools'</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'welch-lab/liger'</span><span class="op">)</span></span></code></pre>
<p>Note that the GitHub version requires installing from source, which may involve additional installation steps on MacOS (see below).</p>
<div class="section level3">
<h3 id="additional-steps-for-installing-liger-from-source-recommended-before-step-3">Additional Steps for Installing LIGER from Source (recommended before step 3)<a class="anchor" aria-label="anchor" href="#additional-steps-for-installing-liger-from-source-recommended-before-step-3"></a>
</h3>
<p>Installation from CRAN is easy because pre-compiled binaries are available for Windows and MacOS. However, a few additional steps are required to install from source on MacOS/Windows (e.g. Install RcppArmadillo). (MacOS) Installing RcppArmadillo on R&gt;=3.4 requires Clang &gt;= 4 and gfortran-6.1. For newer versions of R (R&gt;=3.5), it’s recommended to follow the instructions in this <a href="https://thecoatlessprofessor.com/programming/r-compiler-tools-for-rcpp-on-macos/" class="external-link">post</a>. Follow the instructions below if you have R version 3.4.0-3.4.4.</p>
<ol style="list-style-type: decimal">
<li>Install gfortran as suggested <a href="https://gcc.gnu.org/wiki/GFortranBinaries" class="external-link">here</a>
</li>
<li>Download clang4 from this <a href="https://mac.R-project.org/libs/clang-4.0.0-darwin15.6-Release.tar.gz" class="external-link">page</a>
</li>
<li>Uncompress the resulting zip file and type into Terminal (<code>sudo</code> if needed):</li>
</ol>
<pre><code>mv /path/to/clang4/ /usr/local/</code></pre>
<ol start="4" style="list-style-type: decimal">
<li>Create <code>.R/Makevars</code> file containing following:</li>
</ol>
<pre><code># The following statements are required to use the clang4 binary
CC=/usr/local/clang4/bin/clang
CXX=/usr/local/clang4/bin/clang++
CXX11=/usr/local/clang4/bin/clang++
CXX14=/usr/local/clang4/bin/clang++
CXX17=/usr/local/clang4/bin/clang++
CXX1X=/usr/local/clang4/bin/clang++
LDFLAGS=-L/usr/local/clang4/lib</code></pre>
<p>For example, use the following Terminal commands:</p>
<pre><code>cd ~
mkdir .R
cd .R
nano Makevars</code></pre>
<p>Paste in the required text above and save with <code>Ctrl-X</code>.</p>
</div>
<div class="section level3">
<h3 id="additional-installation-steps-for-online-learning-using-liger">Additional Installation Steps for Online Learning using LIGER<a class="anchor" aria-label="anchor" href="#additional-installation-steps-for-online-learning-using-liger"></a>
</h3>
<p>The HDF5 library is required for implementing online learning in LIGER on data files in HDF5 format. It can be installed via one of the following commands:</p>
<table class="table">
<colgroup>
<col width="55%">
<col width="44%">
</colgroup>
<thead><tr class="header">
<th align="left">System</th>
<th align="left">Command</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left"><strong>OS X (using Homebrew or Conda)</strong></td>
<td align="left">
<code>brew install hdf5</code> or <code>conda install -c anaconda hdf5</code>
</td>
</tr>
<tr class="even">
<td align="left"><strong>Debian-based systems (including Ubuntu)</strong></td>
<td align="left"><code>sudo apt-get install libhdf5-dev</code></td>
</tr>
<tr class="odd">
<td align="left"><strong>Systems supporting yum and RPMs</strong></td>
<td align="left"><code>sudo yum install hdf5-devel</code></td>
</tr>
</tbody>
</table>
<p>For Windows, the latest HDF5 1.12.0 is available at <a href="https://www.hdfgroup.org/downloads/hdf5/" class="external-link uri">https://www.hdfgroup.org/downloads/hdf5/</a>.</p>
</div>
<div class="section level3">
<h3 id="detailed-instructions-for-fit-sne-installation">Detailed Instructions for FIt-SNE Installation<a class="anchor" aria-label="anchor" href="#detailed-instructions-for-fit-sne-installation"></a>
</h3>
<p>Note that the runUMAP function (which calls the <code>uwot</code> package) also scales to large datasets and does not require additional installation steps. However, using FIt-SNE is recommended for computational efficiency if you want to perform t-SNE on very large datasets. Installing and compiling the necessary software requires the use of git, FIt-SNE, and FFTW. For a basic overview of installation, visit this <a href="https://github.com/KlugerLab/FIt-SNE" class="external-link">page</a>.</p>
<p>Basic installation for most Unix machines can be achieved with the following commands after downloading the latest version of FFTW from <a href="http://www.fftw.org/" class="external-link">here</a>. In the fftw directory, run:</p>
<pre><code>./configure
make
make install</code></pre>
<p>(Additional <a href="http://www.fftw.org/fftw3_doc/Installation-and-Customization.html" class="external-link">instructions</a> if necessary). Then in desired directory:</p>
<pre><code>git clone https://github.com/KlugerLab/FIt-SNE.git
cd FIt-SNE
g++ -std=c++11 -O3  src/sptree.cpp src/tsne.cpp src/nbodyfft.cpp  -o bin/fast_tsne -pthread -lfftw3 -lm
pwd</code></pre>
<p>Use the output of <code>pwd</code> as the <code>fitsne.path</code> parameter in runTSNE.</p>
<p>Note that the above instructions require root access. To install into a specified folder (such as your home directory) on a server, use the <code>--prefix</code> option:</p>
<pre><code>./configure --prefix=&lt;install_dir&gt;
make
make install
git clone https://github.com/KlugerLab/FIt-SNE.git
cd FIt-SNE
g++ -std=c++11 -O3  src/sptree.cpp src/tsne.cpp src/nbodyfft.cpp  -I&lt;install_dir&gt;/include/ -L&lt;install_dir&gt;/lib/ -o bin/fast_tsne -pthread -lfftw3 -lm
pwd</code></pre>
</div>
<div class="section level3">
<h3 id="install-time-and-expected-run-time">Install Time and Expected Run Time<a class="anchor" aria-label="anchor" href="#install-time-and-expected-run-time"></a>
</h3>
<p>The installation process of <code>rliger</code> should take less than 30 minutes.</p>
<p>The expected run time is 1 - 4 hours depending on dataset size and downstream analysis of the user’s choice.</p>
</div>
</div>
<div class="section level2">
<h2 id="sample-datasets">Sample Datasets<a class="anchor" aria-label="anchor" href="#sample-datasets"></a>
</h2>
<p>The <code>rliger</code> package provides a small sample dataset for basic demos of the functions. You can find it in folder <code>liger/tests/testdata/small_pbmc_data.RDS</code>.</p>
<p>We also provide a set of scRNA-seq and scATAC-seq datasets for real-world style demos. These datasets are as follows:</p>
<ul>
<li>scRNA data from control and interferon-stimulated PBMCs. Raw data provided by <a href="https://www.nature.com/articles/nbt.4042" class="external-link">Kang, et.al., 2017</a>; The datasets were downsampled by applying the sample function without replacement yield 3000 cells for each matrix. You can download downsampled data from <a href="https://www.dropbox.com/sh/u94ib3dkf9pb6nd/AABemvnxDgKDGRs8Ek5QGlXWa?dl=0" class="external-link">here</a>:
<ul>
<li>
<code>PBMC_control.RDS</code>;</li>
<li>
<code>PBMC_interferon-stimulated.RDS</code>.</li>
</ul>
</li>
<li>scRNA data composed of two datasets of interneurons and oligodendrocytes from the mouse frontal cortex, two distinct cell types that should not align if integrated. Provided by <a href="https://doi.org/10.1016/j.cell.2018.07.028" class="external-link">Saunders, A. et.al., 2018</a>; you can access the pre-processed data from <a href="https://www.dropbox.com/sh/u94ib3dkf9pb6nd/AABemvnxDgKDGRs8Ek5QGlXWa?dl=0" class="external-link">here</a>:
<ul>
<li>
<code>interneurons_and_oligo.RDS</code>;</li>
</ul>
</li>
<li>scATAC and scRNA data provided by <a href="https://github.com/GreenleafLab/MPAL-Single-Cell-2019" class="external-link">GreenleafLab</a>; you can access the pre-processed data from <a href="https://www.dropbox.com/sh/5e9cy4qabs89kc8/AADy3jDxHx94j6A57t7A42u3a?dl=0" class="external-link">here</a>:
<ul>
<li>
<code>GSM4138872_scRNA_BMMC_D1T1.RDS</code>;</li>
<li>
<code>GSM4138873_scRNA_BMMC_D1T2.RDS</code>;</li>
<li>
<code>GSM4138888_scATAC_BMMC_D5T1_peak_counts.RDS</code>;</li>
<li>
<code>GSM4138888_scATAC_BMMC_D5T1.RDS</code>.</li>
</ul>
</li>
</ul>
<!-- * scATAC and scRNA data provided by 10X Genomics, access the pre-processed data from [here](https://umich.app.box.com/s/5hkhpou4inrulo570yhc38ay8g3ont5b). The data sources are: --><ul>
<li>scATAC and scRNA data provided by 10X Genomics. The data sources are:
<ul>
<li>
<code>pbmc.atac.expression.mat.RDS</code>: raw data can be accessed <a href="https://support.10xgenomics.com/single-cell-atac/datasets/1.1.0/atac_v1_pbmc_10k" class="external-link">here</a>, created by Cell Ranger ATAC 1.1.0;</li>
<li>
<code>pbmc.rna.expression.mat.RDS</code>: raw data can be accessed <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/pbmc_10k_v3" class="external-link">here</a>, created by Cell Ranger 3.0.0.</li>
</ul>
</li>
</ul>
<p>Corresponding tutorials can be found in section <strong>Usage</strong> above.</p>
</div>
<div class="section level2">
<h2 id="license">License<a class="anchor" aria-label="anchor" href="#license"></a>
</h2>
<p>This project is covered under the <strong>GNU General Public License 3.0</strong>.</p>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/welch-lab/liger/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/welch-lab/liger/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a> | file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing rliger2</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Joshua Welch <br><small class="roles"> Author </small>  </li>
<li>Yichen Wang <br><small class="roles"> Author, maintainer </small>  </li>
<li>Chao Gao <br><small class="roles"> Author </small>  </li>
<li>Jialin Liu <br><small class="roles"> Author </small>  </li>
<li>Joshua Sodicoff <br><small class="roles"> Author, contributor </small>  </li>
<li>Velina Kozareva <br><small class="roles"> Author, contributor </small>  </li>
<li>Evan Macosko <br><small class="roles"> Author, contributor </small>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
