<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rliger">
<title>UINMF integration of Dual-omics data • rliger</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="UINMF integration of Dual-omics data">
<meta property="og:description" content="rliger">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    <a class="dropdown-item" href="../articles/benchmark.html">Benchmarking integration quality</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>UINMF integration of Dual-omics data</h1>
                        <h4 data-toc-skip class="author">April Kriebel
and Joshua Welch</h4>
            
            <h4 data-toc-skip class="date">04/25/2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/vignettes/articles/SNAREseq_walkthrough.Rmd" class="external-link"><code>vignettes/articles/SNAREseq_walkthrough.Rmd</code></a></small>
      <div class="d-none name"><code>SNAREseq_walkthrough.Rmd</code></div>
    </div>

    
    
<p>Here we integrate the scATAC and scRNA reads from the dual-omics
dataset SNARE-seq as an illustration of how UINMF can be used to
integrate cross-modality data. For this tutorial, we will use three
matrices, which can all be downloaded from our <a href="https://www.dropbox.com/scl/fo/ea666n049h762lla8c5iz/h?rlkey=796slf2ybnv2kmcxxbwluvv2b&amp;dl=0" class="external-link">Dropbox
folder</a></p>
<ul>
<li>The transcriptomic measures <code>SNAREseq_RNA.RDS</code> is the
SNARE-seq scRNA dataset (31,367 genes by 10,309 cells).</li>
<li>For the shared epigenomic features
<code>SNARE_seq_shared_chromatin_features.RDS</code>, we create a
gene-centric matrix, such that we sum of the number of accessibiltiy
peaks that occur over the gene body and promoter regions for each gene.
For a detailed walkthough of how to generate such a matrix, please see
our <a href="Integrating_scRNA_and_scATAC_data.html">Integrating scRNA
and scATAC data vignette</a>. The resulting matrix of gene-centric
chromatin accessibility is 22,379 genes by 10,309 cells</li>
<li>For the unshared epigenomic features, we binned the genome into bins
of 100,000 bp, and summed the number of peaks occuring in each bin. We
then filtered this matrix for all bins that overlapped with ENCODE
Blacklist regions, genes, and promoters. Our filtered matrix
<code>SNARE_seq_unshared_chromatin_features.RDS</code> is 10,309 cells
by 7,437 bins.</li>
</ul>
<div class="section level2">
<h2 id="step-1-load-data">Step 1: Load data<a class="anchor" aria-label="anchor" href="#step-1-load-data"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://welch-lab.github.io/liger/" class="external-link">rliger</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"SNAREseq_data/SNAREseq_RNA.RDS"</span><span class="op">)</span></span>
<span><span class="va">atac_shared</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"SNAREseq_data/SNAREseq_chromatin_accessibility_shared.RDS"</span><span class="op">)</span></span>
<span><span class="va">atac_unshared</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"SNAREseq_data/SNARE_seq_unshared_chromatin_features.RDS"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-2-preprocessing">Step 2: Preprocessing<a class="anchor" aria-label="anchor" href="#step-2-preprocessing"></a>
</h2>
<p>Integrative non-negative matrix factorization with unshared features
(UINMF) performs factorization using shared feature matrix of all
datasets and includes unshared features from dataset(s) with such
information. In this tutorial, we plan to use the variable feature
selected from the scRNA dataset. Since we prepared the gene-centric
matrix from the scATAC dataset, these genes can be accounted as the
shared features. For the unshared features, normally we select variable
genes that are out of the intersection of gene sets from all datasets.
In this tutorial, we directly use the peaks that are identified
variable. Therefore, special processing will be needed compared to the
other UINMF tutorials.</p>
<div class="section level3">
<h3 id="create-liger-object-for-shared-genes">2.1: Create liger object for shared genes<a class="anchor" aria-label="anchor" href="#create-liger-object-for-shared-genes"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createLiger.html">createLiger</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>rna <span class="op">=</span> <span class="va">rna</span>, atac <span class="op">=</span> <span class="va">atac_shared</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="normalize-and-select-shared-variable-features">2.2: Normalize and select shared variable features<a class="anchor" aria-label="anchor" href="#normalize-and-select-shared-variable-features"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalize.html">normalize</a></span><span class="op">(</span><span class="va">lig</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/selectGenes.html">selectGenes</a></span><span class="op">(</span>useDatasets <span class="op">=</span> <span class="st">"rna"</span>, thresh <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="step-3-selecting-the-unshared-features">Step 3: Selecting the unshared features<a class="anchor" aria-label="anchor" href="#step-3-selecting-the-unshared-features"></a>
</h2>
<p>When selecting unshared features for the UINMF integration, it is
critical to consider the type of data you are working with. For unshared
features that gene-centric, the user should follow the feature selection
process outlined in the <a href="UINMF_vignette.html">‘Integrating
unshared features with UINMF’ tutorial</a>.</p>
<p>However, when dealing with features that are <strong>not
gene-centric</strong> (as is the case with our binned intergenic peak
counts), we must select the most variable features using an appropriate
method. Here, we opt to use Seurat’s <code><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures()</a></code>
function.</p>
<p>First, we normalize the unshared features:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">unshareNormed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalize.html">normalize</a></span><span class="op">(</span><span class="va">atac_unshared</span><span class="op">)</span></span></code></pre></div>
<p>Then we select the top 2,000 variable features with Seurat VST
method:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">unshareNormed</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">se</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span>, nfeatures <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="va">top2000</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre></div>
<p>Then we scale, but do not center the unshared features</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">unshareScaled</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="va">unshareNormed</span><span class="op">[</span><span class="va">top2000</span>,<span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>Add the unshared features that have been properly selected, such that
they are added as a genes by cells matrix.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/liger-class.html">varUnsharedFeatures</a></span><span class="op">(</span><span class="va">lig</span>, <span class="st">"atac"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">top2000</span></span>
<span><span class="fu"><a href="../reference/ligerDataset-class.html">scaleUnsharedData</a></span><span class="op">(</span><span class="va">lig</span>, <span class="st">"atac"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">unshareScaled</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-4-joint-matrix-factorization">Step 4: Joint Matrix Factorization<a class="anchor" aria-label="anchor" href="#step-4-joint-matrix-factorization"></a>
</h2>
<p>To factorize the datasets and include the unshared datasets, set
<code>useUnshared = TRUE</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runUINMF.html">runUINMF</a></span><span class="op">(</span><span class="va">lig</span>, k <span class="op">=</span> <span class="fl">30</span>, nIteration <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-5-quantile-normalization-and-joint-clustering">Step 5: Quantile Normalization and Joint Clustering<a class="anchor" aria-label="anchor" href="#step-5-quantile-normalization-and-joint-clustering"></a>
</h2>
<p>After factorization, the resulting Liger object can be used in all
downstream LIGER functions without adjustment. The default reference
dataset for quantile normalization is the larger dataset, but the user
should select the higher quality dataset as the reference dataset, even
if it is the smaller dataset.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantileNorm.html">quantileNorm</a></span><span class="op">(</span><span class="va">lig</span><span class="op">)</span></span>
<span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCluster.html">runCluster</a></span><span class="op">(</span><span class="va">lig</span>, resolution <span class="op">=</span> <span class="fl">0.8</span>, nNeighbors <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-6-visualizations-and-downstream-processing">Step 6: Visualizations and Downstream processing<a class="anchor" aria-label="anchor" href="#step-6-visualizations-and-downstream-processing"></a>
</h2>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">lig</span>, nNeighbors <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<p>Next, we can visualize our returned factorized object by dataset to
check the alignment between datasets, as well as by cluster determined
in the factorization.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>ligerDotSize <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotDimRed.html">plotDatasetDimRed</a></span><span class="op">(</span><span class="va">lig</span><span class="op">)</span></span></code></pre></div>
<p><img src="SNAREseq_walkthrough_files/figure-html/visualizations-1.png" width="672"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDimRed.html">plotClusterDimRed</a></span><span class="op">(</span><span class="va">lig</span>, legendNCol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="SNAREseq_walkthrough_files/figure-html/visualizations-2.png" width="672"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
