<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content='This function aims at calculating the adjusted Rand index for the clustering
result obtained with LIGER and the external clustering (existing "true"
annotation). ARI ranges from 0 to 1, with a score of 0 indicating no
agreement between clusterings and 1 indicating perfect agreement.
The true clustering annotation must be specified as the base line. We suggest
setting it to the object cellMeta so that it can be easily used for many
other visualization and evaluation functions.
The ARI can be calculated for only specified datasets, since true annotation
might not be available for all datasets. Evaluation for only one or a few
datasets can be done by specifying useDatasets. If useDatasets
is specified, the argument checking for trueCluster and
useCluster will be enforced to match the cells in the specified
datasets.'><title>Calculate adjusted Rand index (ARI) by comparing two cluster labeling variables — calcARI • rliger</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.8/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Calculate adjusted Rand index (ARI) by comparing two cluster labeling variables — calcARI"><meta property="og:description" content='This function aims at calculating the adjusted Rand index for the clustering
result obtained with LIGER and the external clustering (existing "true"
annotation). ARI ranges from 0 to 1, with a score of 0 indicating no
agreement between clusterings and 1 indicating perfect agreement.
The true clustering annotation must be specified as the base line. We suggest
setting it to the object cellMeta so that it can be easily used for many
other visualization and evaluation functions.
The ARI can be calculated for only specified datasets, since true annotation
might not be available for all datasets. Evaluation for only one or a few
datasets can be done by specifying useDatasets. If useDatasets
is specified, the argument checking for trueCluster and
useCluster will be enforced to match the cells in the specified
datasets.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Calculate adjusted Rand index (ARI) by comparing two cluster labeling variables</h1>
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/clustering.R" class="external-link"><code>R/clustering.R</code></a></small>
      <div class="d-none name"><code>calcARI.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function aims at calculating the adjusted Rand index for the clustering
result obtained with LIGER and the external clustering (existing "true"
annotation). ARI ranges from 0 to 1, with a score of 0 indicating no
agreement between clusterings and 1 indicating perfect agreement.</p>
<p>The true clustering annotation must be specified as the base line. We suggest
setting it to the object cellMeta so that it can be easily used for many
other visualization and evaluation functions.</p>
<p>The ARI can be calculated for only specified datasets, since true annotation
might not be available for all datasets. Evaluation for only one or a few
datasets can be done by specifying <code>useDatasets</code>. If <code>useDatasets</code>
is specified, the argument checking for <code>trueCluster</code> and
<code>useCluster</code> will be enforced to match the cells in the specified
datasets.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">calcARI</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">trueCluster</span>,</span>
<span>  useCluster <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  useDatasets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span>, <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>  classes.compare <span class="op">=</span> <span class="va">trueCluster</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p>A <a href="liger-class.html">liger</a> object, with the clustering result
present in cellMeta.</p></dd>


<dt>trueCluster</dt>
<dd><p>Either the name of one variable in <code>cellMeta(object)</code>
or a factor object with annotation that matches with all cells being
considered.</p></dd>


<dt>useCluster</dt>
<dd><p>The name of one variable in <code>cellMeta(object)</code>.
Default <code>NULL</code> uses default clusters.</p></dd>


<dt>useDatasets</dt>
<dd><p>A character vector of the names, a numeric or logical
vector of the index of the datasets to be considered for the purity
calculation. Default <code>NULL</code> uses all datasets.</p></dd>


<dt>verbose</dt>
<dd><p>Logical. Whether to show information of the progress. Default
<code>getOption("ligerVerbose")</code> or <code>TRUE</code> if users have not set.</p></dd>


<dt>classes.compare</dt>
<dd><p>[Deprecated/Renamed]. Use <code>trueCluster</code> instead.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A numeric scalar, the ARI of the clustering result indicated by
<code>useCluster</code> compared to <code>trueCluster</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>L. Hubert and P. Arabie (1985) Comparing Partitions, Journal of
the Classification, 2, pp. 193-218.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Assume the true cluster in `pbmcPlot` is "leiden_cluster"</span></span></span>
<span class="r-in"><span><span class="co"># generate fake new labeling</span></span></span>
<span class="r-in"><span><span class="va">fake</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">7</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Insert into cellMeta</span></span></span>
<span class="r-in"><span><span class="va">pbmcPlot</span><span class="op">$</span><span class="va">new</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">fake</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">calcARI</span><span class="op">(</span><span class="va">pbmcPlot</span>, trueCluster <span class="op">=</span> <span class="st">"leiden_cluster"</span>, useCluster <span class="op">=</span> <span class="st">"new"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] -0.0003310172</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Now assume we got existing base line annotation only for "stim" dataset</span></span></span>
<span class="r-in"><span><span class="va">nStim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu"><a href="liger-class.html">dataset</a></span><span class="op">(</span><span class="va">pbmcPlot</span>, <span class="st">"stim"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">stimTrueLabel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">fake</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">nStim</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Insert into cellMeta</span></span></span>
<span class="r-in"><span><span class="fu"><a href="liger-class.html">cellMeta</a></span><span class="op">(</span><span class="va">pbmcPlot</span>, <span class="st">"stim_true_label"</span>, useDatasets <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">stimTrueLabel</span></span></span>
<span class="r-in"><span><span class="co"># Assume "leiden_cluster" is the clustering result we got and need to be</span></span></span>
<span class="r-in"><span><span class="co"># evaluated</span></span></span>
<span class="r-in"><span><span class="fu">calcARI</span><span class="op">(</span><span class="va">pbmcPlot</span>, trueCluster <span class="op">=</span> <span class="st">"stim_true_label"</span>,</span></span>
<span class="r-in"><span>        useCluster <span class="op">=</span> <span class="st">"leiden_cluster"</span>, useDatasets <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0003172964</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

